pkg/controller/rollout/rollout_canary.go

```diff
@@ -363,43 +363,60 @@ func (m *canaryReleaseManager) doCanaryJump(c *RolloutContext) (jumped bool) {
-	if c.NewStatus.CanaryStatus == nil {
+	if canaryStatus == nil {
 		return true, nil
 	}
-	// 1. rollout progressing complete, remove rollout progressing annotation in workload
+	// rollout progressing complete, remove rollout progressing annotation in workload
 	err := m.removeRolloutProgressingAnnotation(c)
 	if err != nil {
 		return false, err
 	}
 	tr := newTrafficRoutingContext(c)
-	// 2. remove stable service the pod revision selector, so stable service will be selector all version pods.
-	done, err := m.trafficRoutingManager.FinalisingTrafficRouting(tr)
-	c.NewStatus.CanaryStatus.LastUpdateTime = tr.LastUpdateTime
-	if err != nil || !done {
-		return done, err
-	}
-	// 3. set workload.pause=false; set workload.partition=0
-	done, err = m.finalizingBatchRelease(c)
-	if err != nil || !done {
-		return done, err
-	}
-	// 4. modify network api(ingress or gateway api) configuration, and route 100% traffic to stable pods.
-	done, err = m.trafficRoutingManager.FinalisingTrafficRouting(tr)
-	c.NewStatus.CanaryStatus.LastUpdateTime = tr.LastUpdateTime
-	if err != nil || !done {
-		return done, err
-	}
-	// 5. delete batchRelease crd
-	done, err = m.removeBatchRelease(c)
-	if err != nil {
-		klog.Errorf("rollout(%s/%s) Finalize batchRelease failed: %s", c.Rollout.Namespace, c.Rollout.Name, err.Error())
+	// execute steps based on the predefined order for each reason
+	nextStep := nextTask(c.FinalizeReason, canaryStatus.FinalisingStep)
+	// if current step is empty, set it with the first step
+	// if current step is end, we just return
+	if len(canaryStatus.FinalisingStep) == 0 {
+		canaryStatus.FinalisingStep = nextStep
+		canaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+	} else if canaryStatus.FinalisingStep == v1beta1.FinalisingStepTypeEnd {
+		klog.Infof("rollout(%s/%s) finalising process is already completed", c.Rollout.Namespace, c.Rollout.Name)
+		return true, nil
+	}
+	klog.Infof("rollout(%s/%s) Finalising Step is %s", c.Rollout.Namespace, c.Rollout.Name, canaryStatus.FinalisingStep)
+
+	var retry bool
+	// the order of steps is maitained by calculating thenextStep
+	switch canaryStatus.FinalisingStep {
+	// set workload.pause=false; set workload.partition=0
+	case v1beta1.FinalisingStepTypeBatchRelease:
+		retry, err = m.finalizingBatchRelease(c)
+	// delete batchRelease
+	case v1beta1.FinalisingStepTypeDeleteBR:
+		retry, err = m.removeBatchRelease(c)
+	// restore the gateway resources (ingress/gatewayAPI/Istio), that means
+	// only stable Service will accept the traffic
+	case v1beta1.FinalisingStepTypeGateway:
+		retry, err = m.trafficRoutingManager.RestoreGateway(tr)
+	// restore the stable service
+	case v1beta1.FinalisingStepTypeStableService:
+		retry, err = m.trafficRoutingManager.RestoreStableService(tr)
+	// remove canary service
+	case v1beta1.FinalisingStepTypeRemoveCanaryService:
+		retry, err = m.trafficRoutingManager.RemoveCanaryService(tr)
+	}
```

Comment:
add a default case to catch potential error

Reply:
done